order.go

```diff
@@ -192,6 +212,59 @@ func (op *OrderPayParams) SetSource(sp interface{}) error {
 	return err
 }
 
+// UnmarshalJSON handles deserialization of an OrderItem.
+// This custom unmarshaling is needed because the resulting
+// Parent property may be an id or a full SKU struct when parent is expanded.
+func (oi *OrderItem) UnmarshalJSON(data []byte) error {
```

Comment:
Thanks for the suggestion! Just to make sure I'm not overlooking something, were you suggesting to take an approach similar to this:
- https://git.corp.stripe.com/gist/spastorelli/9dcfad8c53f4a11054a036ac81dc02b8#file-order-go-L52-L57
- https://git.corp.stripe.com/gist/spastorelli/9dcfad8c53f4a11054a036ac81dc02b8#file-order-go-L218-L245

One issue I would see with moving the custom `UnmarshalJSON` method from `OrderItem` to `OrderItemParent` is that the `data` byte array provided to the call would only include the bytes for the `"parent": ....` property of the `order_item` object [1].

As this `parent` value isn't consistently returned as a subhash, it doesn't always contain an `object` property and as a result `Type` on `OrderItemParent` may not get consistently populated [1] and from my understanding we wouldn't be able to backfill it from the `OrderItem` since it isn't accessible from `OrderItemParent`.

My concern would be that if we provide a `Type` on `OrderItemParent` there could be expectation at the call site that this value is populated.

It may not be an issue but wanted to double check with you just in case.
Let me know what you think.

[1]
For SKU order items, `parent` would be expanded nicely:
```
{
      object: "order_item",
      amount: 499,
      currency: "usd",
      description: "test name",
      parent: {
        id: "TEST-SKU-qSvWbSxdyAHQxEje",
        object: "sku",
        ...
        updated: 1535639219
      },
      quantity: 1,
      type: "sku"
}
```
But for items like `shipping` or `discount` where `parent` doesn't expand (despite requesting to expand it) only the ID string is returned:
```
{
      object: "order_item",
      amount: -399,
      currency: "usd",
      description: "50usdoffonce2: $50.00 off",
      parent: "50usdoffonce2",
      quantity: null,
      type: "discount"
}
```

Reply:
> Thanks for the suggestion! Just to make sure I'm not overlooking something, were you suggesting to take an approach similar to this:

Yeah, that's exactly it!

> One issue I would see with moving the custom `UnmarshalJSON` method from `OrderItem` to `OrderItemParent` is that the `data` byte array provided to the call would only include the bytes for the `"parent": ....` property of the `order_item` object [1].

Ah, I see why you went the way you did.

My preference would still be to change this. I think your justification makes sense, but it feels like we're trying to paper over a broken API with this approach rather than fix the source.

I was just looking at the backend code, and although apparently only SKUs are expandable according to your implementation, it looks at least like the intent was that discounts should be expandable as well, and it's bug if they're not expanding correctly. The lack of expansion on other types is hopefully an oversight that we can correct — ideally, if one type is expandable, than they all are.

While waiting for those fixes though, I don't think it's that bad to just have users check the `Type` on `OrderItem` because they'll be holding a reference to that anyway. e.g.:

``` go
if orderItem.Parent == OrderItemParentTypeSKU {
    // Do something with orderItem.Parent.SKU
```

This keeps the implementation more symmetric with how the API actually looks, and we can expand it if more parents become expandable in the future. What do you think?

ptal @spastorelli-stripe

Reply:
Thanks. Makes sense. Changed as per suggestions. PTAL